storage/blockdevice/COMPONENT_SPINAND/source/SPINANDBlockDevice.cpp

```diff
@@ -0,0 +1,872 @@
+        tr_debug("SPINANDBlockDevice csel: %d", (int)_csel);
+    } else if (_unique_device_status == -1) {
+        tr_error("SPINANDBlockDevice with the same csel(%d) already exists", (int)_csel);
+        return SPINAND_BD_ERROR_DEVICE_NOT_UNIQUE;
+    } else {
+        tr_error("Too many different SPINANDBlockDevice devices - max allowed: %d", SPINAND_MAX_ACTIVE_FLASH_DEVICES);
+        return SPINAND_BD_ERROR_DEVICE_MAX_EXCEED;
+    }
+
+    _mutex.lock();
+
+    // All commands other than Read and RSFDP use default 1-1-1 bus mode (Program/Erase are constrained by flash memory performance more than bus performance)
+    if (QSPI_STATUS_OK != _qspi.configure_format(QSPI_CFG_BUS_SINGLE, QSPI_CFG_BUS_SINGLE, _address_size, QSPI_CFG_BUS_SINGLE,
+                                                 0, QSPI_CFG_BUS_SINGLE, 0)) {
+        tr_error("_qspi_configure_format failed");
+        status = SPINAND_BD_ERROR_DEVICE_ERROR;
+        goto exit_point;
+    }
+
+    if (!_is_initialized) {
+        _init_ref_count = 0;
+    }
+
+    _init_ref_count++;
+
+    if (_init_ref_count != 1) {
+        goto exit_point;
+    }
+
+    _alt_size = 0;
+    _dummy_cycles = 8;
+    if (QSPI_STATUS_OK != _qspi_set_frequency(_freq)) {
+        tr_error("QSPI Set Frequency Failed");
+        status = SPINAND_BD_ERROR_DEVICE_ERROR;
+        goto exit_point;
+    }
+
+    // Synchronize Device
+    if (false == _is_mem_ready()) {
+        tr_error("Init - _is_mem_ready Failed");
+        status = SPINAND_BD_ERROR_READY_FAILED;
+        goto exit_point;
+    }
+
+    if (0 != _clear_block_protection()) {
+        tr_error("Init - clearing block protection failed");
+        status = SPINAND_BD_ERROR_PARSING_FAILED;
+        goto exit_point;
+    }
+
+    if (_read_instruction == SPINAND_INST_READ_CACHE4) {
+        if (QSPI_STATUS_OK != _set_quad_enable()) {
+            tr_error("SPI NAND Set Quad enable Failed");
+            status = SPINAND_BD_ERROR_DEVICE_ERROR;
+            goto exit_point;
+        }
+    }
+
+    _is_initialized = true;
+
+exit_point:
+    _mutex.unlock();
+
+    return status;
+}
+
+int SPINANDBlockDevice::deinit()
+{
+    int result = SPINAND_BD_ERROR_OK;
+
+    _mutex.lock();
+
+    if (!_is_initialized) {
+        _init_ref_count = 0;
+        _mutex.unlock();
+        return result;
+    }
+
+    _init_ref_count--;
+
+    if (_init_ref_count) {
+        _mutex.unlock();
+        return result;
+    }
+
+    // Disable Device for Writing
+    qspi_status_t status = _qspi_send_general_command(SPINAND_INST_WRDI, QSPI_NO_ADDRESS_COMMAND, NULL, 0, NULL, 0);
+    if (status != QSPI_STATUS_OK)  {
+        tr_error("Write Disable failed");
+        result = SPINAND_BD_ERROR_DEVICE_ERROR;
+    }
+
+    _is_initialized = false;
+
+    _mutex.unlock();
+
+    if (_unique_device_status == 0) {
+        remove_csel_instance(_csel);
+    }
+
+    return result;
+}
+
+int SPINANDBlockDevice::read(void *buffer, bd_addr_t addr, bd_size_t size)
```

Comment:
Should `SPINANDBlockDevice::read()` call `nftl_flash_read()` instead of going through QSPI directly? This would allow us to use multiple different flash memory times with littlefs compared to the `#if` approach currently in this PR.

The flow would be something like:
littlefs-> SPINANDBlockDevice::read()-> nftl_flash_read-> QSPIFBlockDevice::read()

SPINANDBlockDevice would act as a virtual block device layer built atop NFTL.